src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,596 @@
+    hasCompleted = false;
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    executionMicroseconds = 0;
+    #endif
+}
+
+CCoinsSelectionAlgorithmBase::~CCoinsSelectionAlgorithmBase()
+{
+    StopSolving();
+}
+
+std::vector<CAmount> CCoinsSelectionAlgorithmBase::PrepareAmounts(std::vector<std::pair<CAmount, size_t>>& amountsAndSizes)
+{
+    std::sort(amountsAndSizes.begin(), amountsAndSizes.end(), std::greater<>());
+    std::vector<CAmount> sortedAmounts(amountsAndSizes.size(), 0);
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        sortedAmounts[index] = amountsAndSizes[index].first;
+    }
+    return sortedAmounts;
+}
+
+std::vector<size_t> CCoinsSelectionAlgorithmBase::PrepareSizes(std::vector<std::pair<CAmount, size_t>>& amountsAndSizes)
+{
+    std::vector<size_t> sortedSizes(amountsAndSizes.size(), 0);
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        sortedSizes[index] = amountsAndSizes[index].second;
+    }
+    return sortedSizes;
+}
+
+bool CCoinsSelectionAlgorithmBase::Reset()
+{
+    bool resetActuallyDone = false;
+
+    if (isRunning || hasCompleted)
+    {
+        if (isRunning)
+        {
+            StopSolving();
+        }
+        for (int index = 0; index < problemDimension; ++index)
+        {
+            tempSelection[index] = 0;
+            optimalSelection[index] = 0;
+        }
+
+        optimalTotalAmount = 0;
+        optimalTotalSize = 0;
+        optimalTotalSelection= 0;
+
+        isRunning = false;
+        stopRequested = false;
+        solvingThread = nullptr;
+        hasCompleted = false;
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        executionMicroseconds = 0;
+        #endif
+        resetActuallyDone = true;
+    }
+
+    return resetActuallyDone;
+}
+
+std::string CCoinsSelectionAlgorithmBase::ToString()
+{
+    return std::string("Input:")
+                                + "{targetAmount=" + std::to_string(targetAmount) +
+                                  ",targetAmountPlusOffset=" + std::to_string(targetAmountPlusOffset) +
+                                  ",availableTotalSize=" + std::to_string(availableTotalSize) + "}\n" +
+                       "Output:"
+                                + "{optimalTotalAmount=" + std::to_string(optimalTotalAmount) +
+                                  ",optimalTotalSize=" + std::to_string(optimalTotalSize) +
+                                  ",optimalTotalSelection=" + std::to_string(optimalTotalSelection) + "}\n";
+}
+
+void CCoinsSelectionAlgorithmBase::StartSolvingAsync()
+{
+    if (!isRunning)
+    {
+        solvingThread = std::unique_ptr<std::thread>(new std::thread(&CCoinsSelectionAlgorithmBase::Solve, this));
+    }
+}
+
+void CCoinsSelectionAlgorithmBase::StopSolving()
+{
+    if (!stopRequested)
+    {
+        stopRequested = true;
+        if (solvingThread != nullptr)
+        {
+            solvingThread->join();
+            solvingThread = nullptr;
+        }
+    }
```

Comment:
Here stopRequested is set to true regardless whether there's an active thread which makes no sense.
The correct implementation should be

```
    bool expected = false;
    if (solvingThread && stopRequested.compare_exchange_strong(&expected, true)) {
        solvingThread->join(); // Wait for completion
        solvingThread.reset(); // Reset to nullptr
        stopRequested = false; // We don't need it anymore
    }
```

Reply:
Based on current implementation, the scenario of a call to `CCoinsSelectionAlgorithmBase::StopSolving` without an active thread can happen if there's a need to stop (from another thread) a (synchronous) solving routine after it has started through `CCoinsSelectionAlgorithmBase::Solve` (please note that currently that's not the case, but I'd prefer to handle that scenario for possible future use).

In addition, the value of `stopRequested` should remain `true` after the actual stop beacuase it is used at the end of `CCoinsSelectionAlgorithmBase::Solve` method (`hasCompleted = !stopRequested;`) to distinguish between a solving routine that reached end due to stop request and a solving routine that reached end due to actual finding completion.

Putting together your suggestion and what I have written above, maybe the best implementation could be:
``
void CCoinsSelectionAlgorithmBase::StopSolving()
{
    if (!stopRequested && isRunning)
    {
        stopRequested = true;
        if (solvingThread != nullptr)
        {
            solvingThread->join();
            solvingThread.reset();
        }
    }
}
``

Reply:
Based on current implementation, the scenario of a call to `CCoinsSelectionAlgorithmBase::StopSolving` without an active thread can happen if there's a need to stop (from another thread) a (synchronous) solving routine after it has started through `CCoinsSelectionAlgorithmBase::Solve` (please note that currently that's not the case, but I'd prefer to handle that scenario for possible future use).

In addition, the value of `stopRequested` should remain `true` after the actual stop beacuase it is used at the end of `CCoinsSelectionAlgorithmBase::Solve` method (`hasCompleted = !stopRequested;`) to distinguish between a solving routine that reached end due to stop request and a solving routine that reached end due to actual completion of optimal searching.

Putting together your suggestion and what I have written above, maybe the best implementation could be:

```
void CCoinsSelectionAlgorithmBase::StopSolving()
{
    if (!stopRequested && isRunning)
    {
        stopRequested = true;
        if (solvingThread != nullptr)
        {
            solvingThread->join();
            solvingThread.reset();
        }
    }
}
```

Reply:
Simplified a little bit with:

```
void CCoinsSelectionAlgorithmBase::StopSolving()
{
    stopRequested = true;
    if (solvingThread != nullptr)
    {
        solvingThread->join();
        solvingThread.reset();
    }
}
```

Reply:
> Based on current implementation, the scenario of a call to CCoinsSelectionAlgorithmBase::StopSolving without an active thread can happen if there's a need to stop (from another thread) a (synchronous) solving routine after it has started through CCoinsSelectionAlgorithmBase::Solve (please note that currently that's not the case, but I'd prefer to handle that scenario for possible future use).

Should this be the case then the whole design is even more entangled and critical :
- either the Solve() method is threaded in both derived classes or it isn't
- if you're thinking "in the future" about possibly stopping `CCoinsSelectionSlidingWindow::Solve()` from a thread which has not started it then you need to redesign the whole base class (and derived ones to be thread-safe: all members exposed with getters must be protected with mutexes and locks on reads and writes otherwise you end up requesting ASYNCRONOUSLY a stop and immediately after you could call `GetOptimalSelection` which returns a vector whose content is totally undefined in lack of memory syncronization.